{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "5af5f63f-3a8e-45d8-8160-969680fc385b",
   "metadata": {},
   "source": [
    "# The `GeneNeighborhood` API\n",
    "\n",
    "## ...And a First Cut at `DuckDB` Database\n",
    "\n",
    "*Mark Green - 3/11/2023*\n",
    "\n",
    "This notebook outlines the `GeneNeighborhood` class. With the DIOPT-provided Gene Homology and identification data stored in the `DuckDB`, this class calls the database to get the neighborhood of immediately connected nodes (genes) to the target geneID. The class builds a `networkx` graph object, and also a `numpy` weighted adjacency matrix. \n",
    "\n",
    "An example call is given, as well as the code for initializing the `DuckDB` - which can be done either in memory or on disk.\n",
    "\n",
    "The `pandas` equivalent is also listed below, for easy reading and experimentation. I don't think we should use this, as I don't believe it is as efficient with data types, nor as flexible future scaling when compared to the `GeneNeighborhood` class. \n",
    "\n",
    "A future extension for this will be to essentially append successive neighborhood graphs. IE, an initial see gene is selected and a neighborhood is built. Then successive genes are \"clicked\" by the user and those gene neighborhoods are fetched and appended to the existing visualized network. Successively expanding the outward, incorporating a new gene homology neighborhood with each \"click\". \n",
    "\n",
    "Additionally, this framework should serve to allow for easy appending of additional node or edge attributes by modifying the `.Graph()` method. For instance, if we want to try analyzing for edge communities. \n",
    "\n",
    "### `../data/` Setup\n",
    "\n",
    "Example setup for the folders, for this script to work. \n",
    "\n",
    "```markdown\n",
    "docs\n",
    "└── duck.db\n",
    "data\n",
    "├── Gene_Information.tsv\n",
    "└── Ortholog_Pair_Best.tsv\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "529e4461-4c9e-416a-8895-39844d6933db",
   "metadata": {},
   "outputs": [],
   "source": [
    "class GeneNeighborhood:\n",
    "    \"\"\"\n",
    "    Generates a local gene homology neighborhood for a given gene\n",
    "    Inputs:\n",
    "        ----::---------------------------------------------------\n",
    "        gid :: int \n",
    "            :: NCBI gene id number\n",
    "        ----::---------------------------------------------------\n",
    "        con :: SQL (DuckDB) db connection\n",
    "            :: 2 tables created from .tsv files provided by DIOPT:\n",
    "            ::    + tblGeneInfo <- 'Gene_Information.tsv'\n",
    "            ::    + tblOrthologPairs <- Ortholog_Pair_Best.tsv\n",
    "        ----::---------------------------------------------------\n",
    "    Dependencies: \n",
    "       + duckdb\n",
    "       + networkx\n",
    "       + numpy\n",
    "       + scipy.sparse\n",
    "    \"\"\"\n",
    "    \n",
    "    def __init__(self, gid, con):\n",
    "        self.gid = gid\n",
    "        self.con = con\n",
    "        self.__CallSQL()\n",
    "        self.__AdjacencyList()\n",
    "    \n",
    "    # API calling SQL db to get data for gene homology neighborhood graph. \n",
    "    def CallSQL(self):\n",
    "        \n",
    "        # drop & build a temp table to list the neighborhood of targetGenes \n",
    "        self.con.execute(\n",
    "            \"\"\"\n",
    "            DROP TABLE IF EXISTS tempTargetGenes; \n",
    "            CREATE TEMP TABLE tempTargetGenes AS \n",
    "                SELECT DISTINCT geneid1 AS geneid FROM tblOrthologPairs \n",
    "                WHERE geneid1 = ? OR geneid2 = ?\n",
    "            \"\"\"\n",
    "            ,[self.gid,self.gid])\n",
    "        \n",
    "        # assign node attribute components\n",
    "        self.targetGenes = self.con.execute(\n",
    "            \"\"\"\n",
    "            SELECT geneid FROM tempTargetGenes ORDER BY geneid\n",
    "            \"\"\"\n",
    "            ).fetchall()\n",
    "        \n",
    "        self.geneInfoNames = self.con.execute(\n",
    "            \"\"\"\n",
    "            SELECT column_name FROM INFORMATION_SCHEMA.COLUMNS \n",
    "            WHERE TABLE_NAME = N'tblGeneInfo'\n",
    "            \"\"\"\n",
    "            ).fetchall()\n",
    "            \n",
    "        self.nodeList = self.con.execute(\n",
    "            \"\"\"\n",
    "            SELECT * FROM tblGeneInfo WHERE geneid IN (\n",
    "                SELECT geneid FROM tempTargetGenes ORDER BY geneid)\n",
    "            \"\"\"\n",
    "            ).fetchall()\n",
    "        \n",
    "        # assign edge attribute components\n",
    "        self.edgeWeights = self.con.execute(\n",
    "            \"\"\"\n",
    "            SELECT geneid1, geneid2, weight FROM tblOrthologPairs \n",
    "            WHERE geneid1 IN (SELECT geneid FROM tempTargetGenes) \n",
    "                AND geneid2 IN (SELECT geneid FROM tempTargetGenes)\n",
    "            \"\"\"\n",
    "            ).fetchall()\n",
    "        \n",
    "        self.orthologPairNames = self.con.execute(\n",
    "            \"\"\"\n",
    "            SELECT column_name FROM INFORMATION_SCHEMA.COLUMNS \n",
    "            WHERE TABLE_NAME = N'tblOrthologPairs'\n",
    "            \"\"\"\n",
    "            ).fetchall()\n",
    "        \n",
    "        self.edgeList = self.con.execute(\n",
    "            \"\"\"\n",
    "            SELECT * FROM tblOrthologPairs \n",
    "            WHERE geneid1 IN (SELECT geneid FROM tempTargetGenes) \n",
    "                AND geneid2 IN (SELECT geneid FROM tempTargetGenes)\n",
    "            \"\"\"\n",
    "            ).fetchall()\n",
    "    \n",
    "    __CallSQL = CallSQL\n",
    "    \n",
    "    # Formats the node and edge attributes to create a NetworkX Graph() object.     \n",
    "    def Graph(self, write=False):\n",
    "\n",
    "        node_attr = dict(zip(\n",
    "            [i for i in range(len(self.targetGenes))],\n",
    "            [dict(zip(\n",
    "                [i for (i,) in self.geneInfoNames],\n",
    "                [i for i in n]\n",
    "            )) for n in self.nodeList]\n",
    "        ))        \n",
    "\n",
    "        edge_attr = np.array([dict(zip(\n",
    "            [i for (i,) in self.orthologPairNames],\n",
    "            [i for i in e]\n",
    "        )) for e in self.edgeList])\n",
    "\n",
    "        graph_edgelist = [tuple(i) for i in np.column_stack((self.adjList[:,:2],edge_attr))]\n",
    "\n",
    "        G = nx.from_edgelist(graph_edgelist)\n",
    "        nx.set_node_attributes(G, node_attr)\n",
    "        \n",
    "        if write:\n",
    "            gexf_path = '../graphs/'+str(self.gid)+'.gexf'\n",
    "            nx.write_gexf(G, gexf_path)\n",
    "            print(f\"{gexf_path} written.\")\n",
    "        \n",
    "        return G\n",
    "    \n",
    "    __Graph = Graph\n",
    "    \n",
    "    # creates weighted edgelist with reindexed node ids.\n",
    "    def AdjacencyList(self):\n",
    "        \n",
    "        node_index_map = dict(zip(\n",
    "            [i for (i,) in self.targetGenes],\n",
    "            [i for i in range(len(self.targetGenes))]\n",
    "        ))\n",
    "        adj_list = np.array(self.edgeWeights).astype(object)\n",
    "        adj_list[:,:2] = np.vectorize(node_index_map.get)(adj_list[:,:2])\n",
    "        self.adjList = adj_list\n",
    "        \n",
    "    __AdjacencyList = AdjacencyList\n",
    "    \n",
    "    # creates weighted adjacency matrix.\n",
    "    def AdjacencyMatrix(self):  \n",
    "        \n",
    "        i, j, weight = self.adjList[:,0], self.adjList[:,1], self.adjList[:,2]\n",
    "        dim = max(len(set(i)), len(set(j)))\n",
    "        adj_mat = sps.lil_matrix((dim, dim))\n",
    "        for i,j,w in zip(i,j,weight):\n",
    "            adj_mat[i,j] = w\n",
    "\n",
    "        return adj_mat.todense()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "2c1306ac-aa16-4974-8bf0-59a7b5915323",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[[0. 1. 1. ... 1. 1. 1.]\n",
      " [1. 0. 4. ... 2. 1. 8.]\n",
      " [1. 4. 0. ... 1. 3. 1.]\n",
      " ...\n",
      " [1. 2. 1. ... 0. 2. 3.]\n",
      " [1. 1. 3. ... 2. 0. 2.]\n",
      " [1. 8. 1. ... 3. 2. 0.]]\n",
      "../graphs/2911.gexf written.\n"
     ]
    }
   ],
   "source": [
    "import duckdb\n",
    "import networkx as nx\n",
    "import numpy as np\n",
    "import scipy.sparse as sps\n",
    "\n",
    "# Initialize db connection\n",
    "con = duckdb.connect(database='duck.db', read_only=True)\n",
    "\n",
    "# ### Alternatively this code could be run to build the dabase in memory from the ./tsv files ###\n",
    "\n",
    "# con = duckdb.connect(database=':memory:')\n",
    "\n",
    "### build db from .tsv sourcefiles\n",
    "# con.execute(\n",
    "#     \"\"\"\n",
    "#     DROP TABLE IF EXISTS tblGeneInfo;\n",
    "#     CREATE TABLE tblGeneInfo AS \n",
    "#         SELECT * FROM read_csv(\n",
    "#             '../data/Gene_Information.tsv', delim='\\t', \n",
    "#             header=True, AUTO_DETECT=TRUE\n",
    "#         );\n",
    "#     \"\"\"\n",
    "# )\n",
    "# ### NOTE: this script should be run in a terminal, may crash Jupyter...\n",
    "# con.execute(\n",
    "#     \"\"\"\n",
    "#     DROP TABLE IF EXISTS tblOrthologPairs;\n",
    "#     CREATE TABLE tblOrthologPairs AS \n",
    "#         SELECT * FROM read_csv(\n",
    "#             '../data/Ortholog_Pair_Best.tsv', delim='\\t', \n",
    "#             header=True, AUTO_DETECT=TRUE\n",
    "#         );\n",
    "#     ALTER TABLE tblOrthologPairs\n",
    "#     RENAME COLUMN score TO weight;\n",
    "#     \"\"\"\n",
    "# )\n",
    "\n",
    "### end alternate ###\n",
    "\n",
    "# Set input ID, visualize graph and write out gexf file for vizualization\n",
    "gid=2911\n",
    "\n",
    "GN = GeneNeighborhood(gid=gid, con=con)\n",
    "\n",
    "print(GN.AdjacencyMatrix())\n",
    "    \n",
    "G = GN.Graph(write=True)\n",
    "\n",
    "# nx.draw(G)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a6d671c3-9a0c-4ab9-bc56-ae9d66469629",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_26212/3933273306.py:8: DtypeWarning: Columns (4) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  gene_info = pd.read_csv('../data/Gene_Information.tsv', sep='\\t')\n"
     ]
    }
   ],
   "source": [
    "import networkx as nx\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "\n",
    "# Here is the pandas version of the above API\n",
    "\n",
    "ortholog_pairs = pd.read_csv('../data/Ortholog_Pair_Best.tsv', sep='\\t')\n",
    "gene_info = pd.read_csv('../data/Gene_Information.tsv', sep='\\t')\n",
    "    \n",
    "# set target geneID\n",
    "gidin = 2911\n",
    "\n",
    "# get ortholog pairs of target gene\n",
    "gidin_orthologs = ortholog_pairs[\n",
    "    (ortholog_pairs['geneid1']==gidin) | \n",
    "    (ortholog_pairs['geneid2']==gidin)\n",
    "]\n",
    "\n",
    "# get ortholog pairs within target gene's ortholog pairs' nodes\n",
    "gidin_all = ortholog_pairs[\n",
    "    (ortholog_pairs['geneid1'].isin(gidin_orthologs['geneid1'])) &\n",
    "    (ortholog_pairs['geneid2'].isin(gidin_orthologs['geneid2']))\n",
    "]\n",
    "\n",
    "# arrange adjacency matrix\n",
    "gidin_adj = pd.crosstab(\n",
    "    gidin_all['geneid1'], gidin_all['geneid2'],\n",
    "    values=gidin_all['score'], aggfunc='sum'\n",
    ")\n",
    "gidin_nodes = gidin_adj.columns.values\n",
    "gidin_mat = gidin_adj.to_numpy()\n",
    "\n",
    "# get node attributes dict\n",
    "gidin_nodes_attr = gene_info[gene_info['geneid'].isin(gidin_nodes)]\\\n",
    "                                                .reset_index(drop=True)\\\n",
    "                                                .to_dict('index')\n",
    "\n",
    "G2 = nx.from_numpy_array(gidin_mat)\n",
    "nx.set_node_attributes(G2, gidin_nodes_attr)\n",
    "\n",
    "# nx.write_gexf(G2, 'GeneNeighborhood2.gexf')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "5ce56ba2-03e7-421e-add3-dcc1ee284b44",
   "metadata": {},
   "outputs": [
    {
     "ename": "TypeError",
     "evalue": "get_node_attributes() missing 1 required positional argument: 'name'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mTypeError\u001b[0m                                 Traceback (most recent call last)",
      "Cell \u001b[0;32mIn [6], line 1\u001b[0m\n\u001b[0;32m----> 1\u001b[0m nx\u001b[38;5;241m.\u001b[39mget_node_attributes(G2)\n",
      "\u001b[0;31mTypeError\u001b[0m: get_node_attributes() missing 1 required positional argument: 'name'"
     ]
    }
   ],
   "source": [
    "nx.get_node_attributes(G2)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4d9f94a0-d58c-4034-b4c9-2a8528ed6b24",
   "metadata": {},
   "source": [
    "A consideration to further increase performance will be to cache the API call results (graphs) in a `redis` or similar database, for users who are clicking through many gene IDs (this will limit I/O pressure from SQL db calls."
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
